files/upsert-records

```diff
@@ -0,0 +1,56 @@
+#!/bin/bash -e
+
+set -o nounset
+umask 177
+
+# Parameters
+url="${1}/api/v1/servers/${2}"
+# $2 is the server
+apikey="${apikey:-}"
+zonename="${3}"
+# The rrset is passed via an env variable because the quotes are mangled when
+# specifying the JSON on the command line.
+expected="${expected:-}"
+
+# Parse zone metadata
+eval "$(echo "${expected}" | jq -r '@sh "defaultTtl=\(.defaultTTL)"')"
+
+# Filter out the records
+records="$(echo "${expected}" | jq '.records')"
+# First step of preparing the RRsets.
+# This groups the records into RRsets.
+# The TTL is at the wrong location, but step 2 fixes that
+# What is left to be done:
+# - Wrap everything into "rrsets"
+# - Add changetype if needed
+records="$(echo "${records}" | jq '[. | to_entries[] | (.value | to_entries[]) as $v | { "name": .key, "type": $v.key, "records": [ $v.value[] | if .c != null then { "content": .c, "disabled": false } + if .r != null then { "set-ptr": .r } else {} end else { "ttl": .t } end ] }]')"
+# Second step moves TTL to correct location and assigns default TTL if needed
+records="$(echo "${records}" | jq --arg defaultTtl "${defaultTtl}" '[.[] | .ttl = (.records | map(select(has("ttl")).ttl))[0] ][] | if .ttl == null then .ttl = ($defaultTtl | tonumber) else .ttl = .ttl end | del(.records[] | select(has("ttl"))) | .name = .name + "."')"
+# Remove PTR and wrap into "rrsets" for comparing with the API
+recordsWithoutPtr="$(echo "${records}" | jq 'del(.records[]."set-ptr")')"
```

Comment:
According to the README, `set-ptr` is not used anymore. Did you mean `r`?
Also, why isn't this done in one line? The rest of this script suggests we don't have enough disk space to be lavish with line breaks.

Reply:
Oh boy. Didn't you read line 26? It sets `set-ptr` there. Also yes, it can be done in one line, but that would increase complexity to a point where I'm neither able to handle nor maintain it.

Reply:
My bad. I searched and marked all occurences of `set-ptr`. I guess the occurence in line 26 was out of vision on my screen because the line is so long.